provider_test.go

```diff
@@ -0,0 +1,150 @@
+package saramakeycloak
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/Nerzal/gocloak"
+	"github.com/stretchr/testify/assert"
+)
+
+const (
+	clientID     = "client-id"
+	clientSecret = "client-secret"
+	realm        = "realm"
+)
+
+func TestProviderGetNewToken(t *testing.T) {
+	p, mock := newProvider(t)
+	defer p.Close()
+	defer mock.MinimockFinish()
+
+	mock.LoginClientMock.
+		Expect(clientID, clientSecret, realm).
+		Return(newJWT("token", 60, "", 0), nil)
+
+	token, err := p.Token()
+
+	assert.NoError(t, err)
+	assert.Equal(t, "token", token.Token)
+}
+
+func TestProviderReturnActiveToken(t *testing.T) {
+	p, mock := newProvider(t)
+	defer p.Close()
+	defer mock.MinimockFinish()
+
+	mock.LoginClientMock.
+		Expect(clientID, clientSecret, realm).
+		Return(newJWT("token", 60, "", 0), nil)
+
+	p.Token()
+	token, err := p.Token()
+
+	assert.NoError(t, err)
+	assert.Equal(t, "token", token.Token)
+	assert.Equal(t, uint64(1), mock.LoginClientAfterCounter())
+}
+
+func TestProviderLoginClientAttemptFails(t *testing.T) {
+	p, mock := newProvider(t)
+	defer p.Close()
+	defer mock.MinimockFinish()
+
+	var calls int
+	mock.LoginClientMock.Set(func(clientID string, clientSecret string, realm string) (jp1 *gocloak.JWT, err error) {
+		if calls == 0 {
+			calls++
+			return nil, errors.New("some error")
+		}
+
+		return newJWT("token", 60, "", 0), nil
+	})
+
+	token, err := p.Token()
+
+	assert.NoError(t, err)
+	assert.Equal(t, "token", token.Token)
+}
+
+func TestProviderRefreshToken(t *testing.T) {
+	p, mock := newProvider(t)
+	defer p.Close()
+	defer mock.MinimockFinish()
+
+	mock.LoginClientMock.
+		Expect(clientID, clientSecret, realm).
+		Return(newJWT("token", 6, "r-token", 0), nil)
+
+	mock.RefreshTokenMock.Expect(
+		"r-token",
+		clientID,
+		clientSecret,
+		realm,
+	).Return(newJWT("token-2", 6, "r-token-2", 10), nil)
+
+	p.Token()
+	time.Sleep(1500 * time.Millisecond)
+	token, err := p.Token()
+
+	assert.NoError(t, err)
+	assert.Equal(t, "token-2", token.Token)
+	assert.Equal(t, uint64(1), mock.LoginClientAfterCounter())
+}
+
+func TestProviderRefreshAttempFails(t *testing.T) {
+	p, mock := newProvider(t)
+	defer p.Close()
+	defer mock.MinimockFinish()
+
+	mock.LoginClientMock.
+		Expect(clientID, clientSecret, realm).
+		Return(newJWT("token", 6, "r-token", 0), nil)
+
+	mock.RefreshTokenMock.
+		Expect("r-token", clientID, clientSecret, realm).
+		Return(nil, errors.New("some error"))
+
+	p.Token()
+	time.Sleep(2000 * time.Millisecond)
+	token, err := p.Token()
+
+	assert.NoError(t, err)
+	assert.Equal(t, "token", token.Token)
+	assert.Equal(t, uint64(2), mock.LoginClientAfterCounter())
```

Comment:
Maybe it makes sense to check calls to `RefreshToken` like `assert.Equal(t, uint64(1), mock.RefreshTokenAfterCounter())`. Maybe not, it's not critical.

Reply:
It's redundant because minimock checks this itself. For `LoginClient` I want to be sure that it was called twice.